webapp/src/main/java/com/box/l10n/mojito/service/ai/openai/OpenAILLMService.java

```diff
@@ -93,6 +117,54 @@ public AICheckResponse executeAIChecks(AICheckRequest aiCheckRequest) {
     return aiCheckResponse;
   }
 
+  @Override
+  @Timed("OpenAILLMService.translate")
+  public String translate(
+      TMTextUnit tmTextUnit, String sourceBcp47Tag, String targetBcp47Tag, AIPrompt prompt) {
+    logger.debug(
+        "Translating text unit {} from {} to {} using prompt {}",
+        tmTextUnit.getId(),
+        sourceBcp47Tag,
+        targetBcp47Tag,
+        prompt.getId());
+    String systemPrompt =
+        getTranslationFormattedPrompt(
+            prompt.getSystemPrompt(), tmTextUnit, sourceBcp47Tag, targetBcp47Tag);
+    String userPrompt =
+        getTranslationFormattedPrompt(
+            prompt.getUserPrompt(), tmTextUnit, sourceBcp47Tag, targetBcp47Tag);
+
+    OpenAIClient.ChatCompletionsRequest chatCompletionsRequest =
+        buildChatCompletionsRequest(
+            prompt, systemPrompt, userPrompt, prompt.getContextMessages(), false);
+
+    return Mono.fromCallable(
+            () -> {
+              OpenAIClient.ChatCompletionsResponse chatCompletionsResponse =
+                  openAIClient.getChatCompletions(chatCompletionsRequest).join();
+              String response = chatCompletionsResponse.choices().getFirst().message().content();
+              logger.debug(
+                  "TmTextUnit id: {}, {} translation response: {}",
+                  tmTextUnit.getId(),
+                  targetBcp47Tag,
+                  response);
+              if (isTranslateResponseJson) {
+                try {
+                  logger.debug("Parsing JSON response for key: {}", translationJsonKey);
+                  response = objectMapper.readTree(response).get(translationJsonKey).asText();
+                  logger.debug("Parsed translation: {}", response);
+                } catch (JsonProcessingException e) {
+                  logger.error("Error parsing JSON response: {}", response, e);
+                  throw new AIException("Error parsing JSON response: " + response);
+                }
+              }
+              return response;
+            })
+        .retryWhen(llmTranslateRetryConfig)
+        .blockOptional()
+        .orElseThrow(() -> new AIException("Error translating text unit " + tmTextUnit.getId()));
```

Comment:
On orElseThrow - could we also capture the underlying reason for the failure?
I.e.: malformed request, rate limiting, service down, 404, etc. ?

Reply:
I can add it to the reactor config to throw the exception if retries are exhausted, in this particular case the optional will be null so there's no information here as to what caused the original failure